Source/Urho3D/Math/Rect.h

```diff
@@ -181,123 +192,53 @@ class URHO3D_API Rect
             return INSIDE;
     }
 
-    /// Return float data.
-    const void* Data() const { return &min_.x_; }
+    /// Return number data.
+    const T* Data() const { return &min_.x_; }
 
     /// Return as a vector.
-    Vector4 ToVector4() const { return Vector4(min_.x_, min_.y_, max_.x_, max_.y_); }
+    BaseVector4<T> ToVector4() const { return BaseVector4<T>(min_.x_, min_.y_, max_.x_, max_.y_); }
 
     /// Return as string.
     String ToString() const;
 
-    /// Minimum vector.
-    Vector2 min_;
-    /// Maximum vector.
-    Vector2 max_;
+    union
+    {
+        /// Minimum vector.
+        BaseVector2<T> min_;
+        struct
+        {
+            /// Left coordinate.
+            T left_;
+            /// Top coordinate.
+            T top_;
+        };
+    };
+
+    union
+    {
+        /// Maximum vector.
+        BaseVector2<T> max_;
+        struct
+        {
+            /// Right coordinate.
+            T right_;
+            /// Bottom coordinate.
+            T bottom_;
+        };
+    };
 
     /// Rect in the range (-1, -1) - (1, 1)
-    static const Rect FULL;
+    static const BaseRect FULL;
     /// Rect in the range (0, 0) - (1, 1)
-    static const Rect POSITIVE;
+    static const BaseRect POSITIVE;
     /// Zero-sized rect.
-    static const Rect ZERO;
+    static const BaseRect ZERO;
 };
 
-/// Two-dimensional bounding rectangle with integer values.
-class URHO3D_API IntRect

```

Comment:
IntRect logic has changed in many places.
1. Default ctor values
2. Clip and Merge conditions and so on

Reply:
Please elaborate.

Reply:
IntRect() was (0, 0, 0, 0), become (2kkk, 2kkk, -2kkk, -2kkk).
So, IntRect is no longer empty, Width and Height are broken. Clip is broken too.
Then, the IntRect has exclusive upper border and Rect has inclusive one. So literally _all_ places with > and < signs are broken for IntRect now. Clip, Merge, IsInside...

Reply:
> IntRect() was (0, 0, 0, 0), become (2kkk, 2kkk, -2kkk, -2kkk).

Is there a good reason for `Rect` to define infinitely large rect? Maybe it should become `(0, 0, 0, 0)` instead? Does infinitly large rect have a purpose?

> Then, the IntRect has exclusive upper border and Rect has inclusive one.
> Clip, Merge, IsInside...

Good spot. Ill add template specialization restoring old behavior to `IntRect`. For the record: one of these rects has a bug. Not sure if we want to fix it though.

Reply:
> Maybe it should become (0, 0, 0, 0) instead?

Rect (0, 0, 0, 0) is considered as valid rect with zero area. Point (0, 0) is inside the rect.
Rect(inf, inf, -inf, -inf) is considered as invalid rect. Point (0, 0) is outside the rect because there is no rect.
So we should keep old behavior here.

> For the record: one of these rects has a bug. Not sure if we want to fix it though.

Huh, what bug do you mean?

Reply:
> Huh, what bug do you mean?

I meant inconsistent `IsInside()` checks, but i guess all of the inconsistent behavior should be called a bug.

Reply:
Rect over discrete space (int) is exclusive, rect over continious space (float) is inclusive. This is the most convinient way of defining rects, so it's not a bug and will never be changed.

In the perfect world, template BaseRect should behave as Rect for real types and as IntRect for integer ones.

Reply:
Got it. I think i can make that happen.